contracts/insurance/Insurance.sol

```diff
@@ -0,0 +1,119 @@
+  string ResolverEncoding = "tuple(tuple(uint256 amount, bytes32 UUID) data, bytes signature)";
+  string Name = "Insurance";
+  /* solhint-enable */
+
+  function getRegistryInformation() external override view returns (RegisteredTransfer memory) {
+    RegisteredTransfer memory info = RegisteredTransfer({
+      name: Name,
+      stateEncoding: StateEncoding,
+      resolverEncoding: ResolverEncoding,
+      definition: address(this)
+    });
+    return info;
+  }
+
+  /// @notice Creates an insurancePayment from the guarantor to the receiver
+  /// @param encodedBalance balance the guarantor is putting into the transfer
+  /// @param encodedState encoded transfer parameters (insurance payment info)
+  /// @dev encoded state info contains the receiver & mediator addresses, collateral amount, expiration, and UUID
+  function create(
+		bytes calldata encodedBalance,
+		bytes calldata encodedState
+	) external override view returns (bool) {
+    Balance memory balance = abi.decode(encodedBalance, (Balance));
+    TransferState memory state = abi.decode(encodedState, (TransferState));
+
+    // Receiver should not have an initial balance within the transfer
+    require(balance.amount[1] == 0, "Cannot create parameterized payment with nonzero recipient init balance");
+
+    // Insurance payment checks
+    require(state.expiration > block.timestamp + 3 days, "Expiration must be at least 3 days in the future.");
+
+    // Sanity checks
+    require(state.receiver != address(0x0), "Receiver address cannot be the zero address!");
+    require(state.mediator != address(0x0), "Mediator address cannot be the zero address!");
+    require(state.UUID != bytes32(0), "UUID cannot be null.");
+    require(state.collateral >= 1 wei, "Collateral must be nonzero");
+
+    return true;
+  }
+
+  /// @notice Resolves an insurance payment with a mediator signature, or cancels it with the receiver signature
+  /// @param encodedBalance balance the guarantor is putting into the tranfser
+  /// @param encodedState encoded transfer parameters (insurance payment info)
+  /// @param encodedResolver contains the resolution data & the mediator signature
+  function resolve(
+    bytes calldata encodedBalance,
+    bytes calldata encodedState,
+    bytes calldata encodedResolver
+  ) external override view returns (Balance memory) {
+    // Unpack balance, state, resolver, & resolverdata
+    Balance memory balance = abi.decode(encodedBalance, (Balance));
+    TransferState memory state = abi.decode(encodedState, (TransferState));
+    TransferResolver memory resolver = abi.decode(encodedResolver, (TransferResolver));
+    ResolverData memory data = resolver.data;
+
+    // State & resolver UUID should match
+    require(state.UUID == data.UUID, "UUID did not match!");
+
+    // Expiration check
+    require(block.timestamp < state.expiration, "Insurance payment expired.");
```

Comment:
Can the payment not be cancelled after it expires? It seems the payment should revert so no balances remain locked after the expiry. Perhaps this could be done with a sender or empty signature?

Would definitely recommend making sure the payment can be removed or reverted before it has been cancelled, though perhaps I'm misunderstanding something about your target use case.

Reply:
Changed this to `require(data.paymentAmountTaken == 0 || block.timestamp < state.expiration, "Payment expired!");`

Reply:
Changed to

```
    // If expired, no payment can happen.
    if (block.timestamp >= state.expiration) {
      return balance;
    }
```